### DISCUSSION TITLE
Is there a way you're supposed to work with the out keyword and nullable types?

### DISCUSSION THREAD

**Discussion by jklw10** [dotnet/csharplang#6936] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/8884828/215948305-61073a72-fc3d-4c06-ad36-d0c8e8ddf20f.png)

As you can see in this example, none of the ways that make sense to me work for getting the possible value out from the MathContext.Calculate()'s out argument.
the way i do it takes many more lines to make all the warnings go away.

![image](https://user-images.githubusercontent.com/8884828/215948576-4f7a082c-a073-4edb-9157-109786a103b2.png)

I shouldn't have to use the ! operator right?

![image](https://user-images.githubusercontent.com/8884828/215948816-96849470-fc77-434d-8072-5846c790e0ae.png)

if i use ? it says cthird is now unassigned and spouts an error. which, it should be either null, or atleast allow you to accompany that out with a null.

![image](https://user-images.githubusercontent.com/8884828/215950162-8ad6438a-87bf-49a1-ba29-5e0c49c9a7be.png)

here's how i see it should look.

**Comment by siegfriedpammer** [comment]
You should be able to use the `NotNullWhenAttribute`:

Example signature:

     public bool MatchLdLoc([NotNullWhen(true)] out ILVariable? variable)

Full example:

https://github.com/icsharpcode/ILSpy/blob/2ef3245158ca6d7001475115845af31855e578a8/ICSharpCode.Decompiler/IL/Instructions.cs#L8192-L8202

**Reply by jklw10** [reply]
Ah a good thing to learn for certain, but might not be possible if I wasn't the one writing the library. and the calculate() output isn't nullable the nullability is only in the MathPieces's components.

**Comment by yaakov-h** [comment]
You appear to be battling with definite assignment, not nullability.

Assuming that `Token` is a reference type, then you can simply call `tokens[2].mathContext?.Calculate(out third2);`. You don't need to use `out var` when you already have a variable that you want to assign the result to.

If `Token` is a value type, then you either need to match the nullability of `third2` with the nullability of your `out` parameter, or you need that `if` block to wrap or unwrap the result.

**Comment by koszeggy** [comment, marked as answer]
Due to null short-circuiting this behavior is expected. If `tokens[2].mathContext` is `null`, then `Calculate` is not executed so your `cthird` remains unassigned. You even highlighted CS0165, which explains the issue.

But since you already initialized `third2` there is no need to introduce another variable in the out parameter and then assign it to `third2`:

```cs
// This causes CS0165 because cthird will not be assigned if mathContext is null:
//Token third2 = Token.Invalid;
//tokens[2].mathContext?.Calculate(out Token cthird);
//third2 = cthird; // issue: cthird might be uninitialized here

// But you can simply do this:
Token third2 = Token.Invalid;
tokens[2].mathContext?.Calculate(out third2); // third2 is now assigned even if Calculate didn't execute
```

ps: and please avoid posting code as image because it's harder for us to answer if we cannot copy-paste it.

**Reply by jklw10** [reply]
```cs
static Token GetToken(MathPiece mp)
{
    if (mp.MathContext is not null)
    {
        mp.MathContext.Calculate(out Token cthird);
        return mp.Token ?? cthird;
    }
    return mp.Token ?? Token.INVALID;
}
static Token GetToken2(MathPiece mp)
{
    Token TokenOut = mp.Token ?? Token.INVALID;
    mp.MathContext?.Calculate(out TokenOut);
    return TokenOut;
}
```
the first one is what I Ended up with yesterday that works, but yeah the ``GetToken2()``'s things could've been inside the the original piece of code without method extraction. which were written based on your suggestion.
